keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,787 @@
+This means that when this feature is turned on, the job controller can count pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.  So this pod would be considered active until it is fully terminated.  
+If `PodDisruptionCondition` is turned off then the job controller considers the pod as failed as soon as it is terminating (has a deletion timestamp), because there is no guarantee that the pod will transition to phase=Failed.
+
+Another issue is described [here](https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509).
+If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.  Due to the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.  We will need to modify the PodGC (gc_controller.go) in the case of `JobRecreatePodsFailed` being enabled while `PodDisruptionConditions` is disabled.  
+
+## Design Details
+
+### Job API Definition
+
+At the JobSpec level, we are adding a new enum field:
+
+```golang
+// This field controls when we recreate pods
+// Default will be TerminatingOrFailed ie recreate pods when they are failed
+// +enum 
+type RecreatePodsWhen string
+const (
+ // This is a field that recreates pods when they are marked as terminating or failed
+ // ie this means that as soon as pods get marked for deletion they will be recreated
+ TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
+ // Only recreate pods when they are marked as failed.
+ Failed              RecreatePodsWhen = "Failed"
+)
+```
+
+```golang
+type JobSpec struct{
+  ...
+ // RecreatePodsWhen specifies when pods should be recreated.
+ // TerminatingOrFailed means to recreate when a pod is either terminating or failed
+ // TerminatingOrFailed is the default.
+ // Failed means to wait until pods are fully terminated or failed before recreating
+ // +optional
+ RecreatePodsWhen *RecreatePodsWhen
+}
+```
+
+So we can count terminating pods separately from active or failed we need to include a new field in the JobStatus.
+
+```golang
+type JobStatus struct {
+  ...
+  // Number of terminating pods
+  // +optional
+  terminating *int32
+}
+```
+
+We will allow only opt-in behavior for this feature so we will fall back to `TerminatingOrFailed` for `RecreatePodsWhen`.  
+
+### Implementation
+
+As part of this KEP, we want to include pods that are also terminating (`DeletionTimestamp != nil`).  
+
+The following algorithm could be used:
+
+1) Count the number of pods that are active
+2) If `Failed` and feature is on, then we will count the number of terminating pods
```

Comment:
Why are terminating pods counted under `Failed` instead of under `TerminatingOrFailed`?  Typo?

Reply:
Not a typo, but we should always calculate terminating pods. Adjusted the algorithm.